### PR TITLE
Allow clean exit in boutcore

### PR DESCRIPTION
**Pull request by dschwoerer**
It's not particularly nice, to duplicating `__dealloc__` - but `fu.__dealloc__()` cannot be called, and simply calling `del fu` doesn't work ether, as long as there are other references.

`atexit` should ensure we are called rather late, to finalise. Any remaining references still need to be manually taken care of.

### CODE REVIEW

File: tools/pylib/_boutcore_build/boutcore.pyx.in [hunk 1]
```diff
@@ -1604,36 +1671,52 @@ cdef class Options:
...
+        cdef c.Field3D * f3d
+        t2=str.encode(name)
+        tmp=t2
+        # import numbers
+        # if isinstance(data,int):
+        #     #c_datafile_add_int(self.cobj,
+        #     self.cobj.add(<int>data,tmp,<c.bool>save_repeat)
+        # elif isinstance(data,numbers.Number):
+        #     self.cobj.add(dbl,tmp,<c.bool>save_repeat)
+        # el
```
Comment (ZedThree) [inline tools/pylib/_boutcore_build/boutcore.pyx.in]: Please don't commit commented out code

File: tools/pylib/_boutcore_build/boutcore.pyx.in [hunk 2]
```diff
@@ -1118,17 +1175,27 @@ def init(args=[]):
...
     Finalize BOUT++ and also MPI.
     After this most objects and functions are not valid anymore.
     """
-    checkInit()
-    BoutFinalise()
     global _isInit
-    _isInit=False
+    import gc
+    ls = gc.get_objects()
+    im=(Coordinates, Field2D, Field3D, Laplacian, Mesh, Options,
```
Comment (ZedThree) [inline tools/pylib/_boutcore_build/boutcore.pyx.in]: Please could you give these more descriptive names? It's hard to work out what's going on

File: tools/pylib/_boutcore_build/boutcore.pyx.in [hunk 3]
```diff
@@ -1118,17 +1175,27 @@ def init(args=[]):
...
+        PhysicsModel, PhysicsModelBase, Vector2D, Vector3D,
+        PythonModelCallback)
+    print("isInit is",_isInit)
```
Comment (ZedThree) [inline tools/pylib/_boutcore_build/boutcore.pyx.in]: Did you mean to leave this `print` in?

File: tools/pylib/_boutcore_build/boutcore.pyx.in (hunk 2 above)
Comment (dschwoerer) [reply tools/pylib/_boutcore_build/boutcore.pyx.in]: I also added some documentation, I hope that helps

**Comment by dschwoerer** [comment]
The [pr tests](https://travis-ci.org/boutproject/BOUT-dev/builds/515617908) are all failing - however I doubt it is related to this PR ...

**Comment by ZedThree** [comment]
Thanks @dschwoerer, that's made the tests much nicer! valgrind is happier now

**Pull request merged**